source/simulator/assemblers/adjoint.cc

```diff
@@ -0,0 +1,121 @@
+
+      const double pressure_scaling = this->get_pressure_scaling();
+
+      // Get a pointer to the dynamic topography postprocessor.
+      Postprocess::DynamicTopography<dim> &dynamic_topography = const_cast<Postprocess::DynamicTopography<dim> &> (
+                                                                  this->get_postprocess_manager().template get_matching_postprocessor<Postprocess::DynamicTopography<dim> >());
+
+      // Get the already computed dynamic topography solution.
+      const LinearAlgebra::BlockVector &topography_vector = dynamic_topography.topography_vector();
+      std::vector<double> topo_values( n_face_q_points );
+
+      // Get the density_above from the dynamic topography postprocessor
+      const double density_above = dynamic_topography.return_density_above();
+
+      // check that the cell is at the top and that the boundary is at the top
+      if (scratch.cell->face(scratch.face_number)->at_boundary()
+          &&
+          scratch.cell->face(scratch.face_number)->boundary_id() == this->get_geometry_model().translate_symbolic_boundary_name_to_id("top"))
+        {
+          // get values at the surface of the cell
+          scratch.face_finite_element_values.reinit (scratch.cell, scratch.face_number);
+          scratch.face_finite_element_values[introspection.extractors.temperature].get_function_values(topography_vector, topo_values);
+
+          // ----------  Assemble RHS  ---------------
+
+          for (unsigned int q=0; q<n_face_q_points; ++q)
+            {
+              for (unsigned int i=0, i_stokes=0; i_stokes<stokes_dofs_per_cell; /*increment at end of loop*/)
+                {
+                  if (introspection.is_stokes_component(fe.system_to_component_index(i).first))
+                    {
+                      scratch.phi_p[i_stokes] = scratch.face_finite_element_values[introspection.extractors.pressure].value (i, q);
+                      scratch.grads_phi_u[i_stokes] = scratch.face_finite_element_values[introspection.extractors.velocities].symmetric_gradient(i,q);
+                      ++i_stokes;
+                    }
+                  ++i;
+                }
+
+
+              const Tensor<1,dim> n_hat = scratch.face_finite_element_values.normal_vector(q);
+              const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector (scratch.face_finite_element_values.quadrature_point(q));
+              const double density = scratch.material_model_outputs.densities[q];
+              const double eta = scratch.material_model_outputs.viscosities[q];
+              const double JxW = scratch.face_finite_element_values.JxW(q);
+
+              for (unsigned int i=0; i<stokes_dofs_per_cell; ++i)
+                {
+                  // The new RHS is the derivative of the objective functional with respect to pressure and velocity
+                  data.local_rhs(i) += topo_values[q] * (2.0*eta *(n_hat * (scratch.grads_phi_u[i] * n_hat))
```

Comment:
Yes that's right. Currently, it calculates the sensitivity to the current topography or in other words it assumes the observed topography is zero. There are benchmarks for this scenario and it's sort of the easiest case. I would suggest leaving this as is for this first pull request but you're right that including observations is the logical next step (and I have done this already on a different branch to test it out).

Reply:
Yes that's right. Currently, it calculates the sensitivity to the current topography or in other words it assumes the observed topography is zero. There are benchmarks for this scenario and it's sort of the easiest case. I would suggest leaving this as is for this first pull request but you're right that including observations is the logical next step (and I have done this already on a different branch to test it out).

Reply:
I left a comment here and in the helper_function to indicate that it currently calculates the gradient with respect fo the squared topography (instead of a misfit).